subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskExecution.java

```diff
@@ -458,17 +458,25 @@ public void ensureLegacySnapshottingInputsClosed() {
 
     @Override
     public void validate(WorkValidationContext validationContext) {
-        Class<?> taskType = GeneratedSubclasses.unpackType(task);
-        // TODO This should probably use the task class info store
-        boolean cacheable = taskType.isAnnotationPresent(CacheableTask.class);
-        TypeValidationContext typeValidationContext = validationContext.forType(taskType, cacheable);
+        TypeValidationContext typeValidationContext = getTypeValidationContext(validationContext);
         context.getTaskProperties().validateType(typeValidationContext);
         context.getTaskProperties().validate(new DefaultPropertyValidationContext(
             fileResolver,
             reservedFileSystemLocationRegistry,
             typeValidationContext
         ));
-        context.getValidationAction().validate(typeValidationContext);
+    }
+
+    @Override
+    public void checkDependencies(WorkValidationContext validationContext) {
+        context.getDependencyCheckAction().checkDependencies(getTypeValidationContext(validationContext));
+    }
```

Comment:
💭 This is the main part of the change: we've separated "validation" (=checking the implementation type and the properties) and "dependency check" (=the missing task dependency check), and the latter is now always executed, while the former is only executed after the "skip-when-empty" check (as it was before).